ENH: Add require argument to load() to accept version specifiers

effigies opened the pull request:
This PR adds an optional `require` keyword argument to `load()` that accepts `requirements.txt`-style version specifiers.

```python
np = lazy.load("numpy", require="numpy >=1.24")
```

The effect is to make the import fail if the dependency exists but the version is not satisfied.

Closes #13.

---- 

<details><summary>Original text</summary>

Found a little time tonight to take a first stab at implementing `lazy.load_requirement()`, a variant of `lazy.load()` that accepts a `requirements.txt`-style requirement specifier. This would allow one to write:

```Python
nibabel, have_nibabel = lazy.load_requirement('nibabel >= 4.2')
yaml, have_yaml = lazy.load_requirement('pyyaml >= 6.0', 'yaml')
stats, have_scipy = lazy.load_requirements('scipy >= 1.9', 'scipy.stats')
```

The two-argument form is for when module names do not match distribution names or you want to lazily import a submodule while constraining the base module version.

I'll get to tests when I can, but might as well see if this much complexity is within-scope. I doubt it can be done much simpler.

</details>

stefanv commented:
I think this could work!

Would we be able to re-use some existing code by making the interface `lazy.load(spec, requires=...)`, instead of defining an entirely new function?

Adding the new function is also fine, although I would swap the order of the arguments `scipy.stats, scipy >= 1.9`.

stefanv commented on lazy_loader/__init__.py:
```diff
@@ -160,38 +166,72 @@ def myfunc():
+    if not have_module:
+        not_found_message = f"No module named '{fullname}'"
+    elif require is not None:
+        # Old style lazy loading to avoid polluting sys.modules
```
What does this comment mean?

effigies replied:
Ah. Initially I'd used `load()` to lazily load `packaging.requirements` at the module root, but it seems better not to add it to `sys.modules` unless calling code actually makes use of `packaging.requirements`. So "old style lazy loading" just means import at function runtime.

stefanv commented on lazy_loader/__init__.py:
```diff
@@ -160,38 +166,72 @@ def myfunc():
+        except importlib_metadata.PackageNotFoundError as e:
+            raise ValueError(
+                f"Found module '{fullname}' but cannot test requirement '{require}'. "
+                "Requirements must match distribution name, not module name."
```
What does this refer to?

effigies replied:
I suspect a common error will be to do something like `yaml = load('yaml', 'yaml >= 5')`. The `yaml` module is provided by the `pyyaml` distribution, so it needs to be `yaml = load('yaml', 'pyyaml >= 5')`.

This is necessary because not all modules provide a `__version__`, but all distributions must provide version metadata, and that's what's retrieved by `importlib.metadata.version`.

stefanv commented on lazy_loader/__init__.py:
```diff
@@ -160,38 +166,72 @@ def myfunc():
     return module
 
 
+def have_module(module_like: types.ModuleType) -> bool:
```
Is `have_module` used anywhere?

effigies replied:
Not internally. I thought it would be a nice addition to the API, as I will often do something like:

```Python
# Was: h5py, have_h5py, _ = optional_package('h5py')
h5py = lazy.load('h5py')
have_h5py = lazy.have_module(h5py)

needs_h5py = unittest.skipUnless(have_h5py, 'Test requires h5py')

@needs_h5py
def test_mything():
    ...
```

I included it here because I initially had `load_requirement() -> tuple[module, boolean]`, so this replaced that second return value. Could be a separate PR to keep this clean, or I can just do it downstream, myself. The lazy loading and deferred errors are the important things to me.

stefanv reviewed (commented):
This looks good already!

effigies commented:
I was curious about how heavy these imports are, so I moved importing `inspect`, `packaging.requirements` and `importlib.metadata` inside the functions that need them.

<details><summary>Click for diff</summary>

```diff
diff --git a/lazy_loader/__init__.py b/lazy_loader/__init__.py
index 016c26d..f3427c1 100644
--- a/lazy_loader/__init__.py
+++ b/lazy_loader/__init__.py
@@ -7,16 +7,10 @@ Makes it easy to load subpackages and functions on demand.
 import ast
 import importlib
 import importlib.util
-import inspect
 import os
 import sys
 import types
 
-try:
-    import importlib_metadata
-except ImportError:
-    import importlib.metadata as importlib_metadata
-
 __all__ = ["attach", "load", "attach_stub"]
 
 
@@ -181,9 +175,14 @@ def load(fullname, *, require=None, error_on_import=False):
     if not have_module:
         not_found_message = f"No module named '{fullname}'"
     elif require is not None:
         # Old style lazy loading to avoid polluting sys.modules
         import packaging.requirements
 
+        try:
+            import importlib_metadata
+        except ImportError:
+            import importlib.metadata as importlib_metadata
+
         req = packaging.requirements.Requirement(require)
         try:
             have_module = req.specifier.contains(
@@ -202,6 +201,8 @@ def load(fullname, *, require=None, error_on_import=False):
     if not have_module:
         if error_on_import:
             raise ModuleNotFoundError(not_found_message)
+        import inspect
+
         try:
             parent = inspect.stack()[1]
             frame_data = {
```

</details>

Then I profiled loading `lazy_loader` and each of these requirements in turn.

```console
$ python -X importtime -c "import lazy_loader; import inspect; import packaging.requirements; import importlib.metadata" 2>&1 | awk '/[0-9e] \| [a-z]/' 
import time: self [us] | cumulative | imported package
[...]
import time:       205 |       4984 | lazy_loader
import time:      1067 |       3710 | inspect
import time:       156 |      19275 | packaging.requirements
import time:       860 |      12247 | importlib.metadata
```

`packaging.requirements` and `importlib.metadata` are each 2-4x slower to load than `lazy_loader` itself, and only needed if this feature is called for. That said, we're talking about a one-time import cost of ~30-35ms. I'll leave it to you to decide if we should import everything at the module level or defer some until needed.

stefanv commented:
I think this is the kind of library where that type of optimization makes perfect sense.

effigies commented:
Rebased, smoothed down ruff's feathers about how complex `load()` was getting, and removed the `have_module()` function as out-of-scope.

This is ready for review, whenever it's convenient.

dschult reviewed (commented):
This looks like a reasonable feature and it seems set up well. I don't know the `mock` unittest construct details but the intent seems good. I put a few comments below as suggestions.

effigies commented on lazy_loader/__init__.py:
```diff
@@ -99,24 +98,25 @@ def __dir__():
-    def __init__(self, frame_data, *args, **kwargs):
+    def __init__(self, frame_data, *args, message, **kwargs):
         self.__frame_data = frame_data
+        self.__message = message
```
> Do you want a default message?

I found it was easier to be more explicit about setting the message in the branching logic. I can rework with a default, if preferred.

> Also, why is message after *args in the function sig instead of before? For backward compat?

I figured keyword-only would be clearer, but I'm okay with any signature. Let me know if you'd like me to change this.

dschult replied:
Yes, I see what you mean and agree that dealing with the message is better handled in the branching logic.

dschult reviewed (approved):
I approve this PR (subject to the latest changes passing tests once the testing workflows  are approved). 
Thanks!

stefanv commented:
Thank you very much @effigies, for your contribution and your patience while we decided how to handle this.
